src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -393,6 +408,42 @@ protected void CheckDocumentIsOpen(DocumentId documentId)
             this.RegisterText(textContainer);
         }
 
+        /// <summary>
+        /// Registers a SourceTextContainer to a source generated document. Unlike <see cref="OnDocumentOpened" />,
+        /// this doesn't result in the workspace being updated any time the contents of the container is changed; instead
+        /// this ensures that features going from the text container to the buffer back to a document get a usable document.
+        /// </summary>
+        // TODO: switch this protected once we have confidence in API shape
+        internal void OnSourceGeneratedDocumentOpened(
+            SourceGeneratedDocumentIdentity documentIdentity,
+            SourceTextContainer textContainer)
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                var documentId = documentIdentity.DocumentId;
+                CheckDocumentIsClosed(documentId);
+                AddToOpenDocumentMap(documentId);
+
+                _documentToAssociatedBufferMap.Add(documentId, textContainer);
+                _openSourceGeneratedDocumentIdentities.Add(documentId, documentIdentity);
+
+                UpdateCurrentContextMapping_NoLock(textContainer, documentId, isCurrentContext: true);
+            }
+
+            this.RegisterText(textContainer);
+        }
+
+        internal void OnSourceGeneratedDocumentClosed(DocumentId documentId)
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                Contract.ThrowIfFalse(_openSourceGeneratedDocumentIdentities.Remove(documentId));
+                CheckDocumentIsOpen(documentId);
+                AddToOpenDocumentMap(documentId);
```

Comment:
why are we adding the doc to the open doc map during close?  at the very least, plz doc. :)

Reply:
Oh that's not good.